### ISSUE TITLE
feat(nn): Multiplicative Filter Networks (FourierNet, GaborNet)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#87]
## Summary

Add Multiplicative Filter Networks (MFN) to `pyrox.nn` — both the primitive filters (`FourierFilter`, `GaborFilter`) and two ready-to-use composite networks (`FourierNet`, `GaborNet`) that chain filters multiplicatively. MFN is the principled alternative to SIREN for implicit neural representations: a network that is a *product of filters of the input* rather than a deep composition of nonlinearities, with demonstrably better conditioning, no special initialization ceremony, and a direct mathematical bridge to the spectral-kernel machinery already in pyrox.

Landing this gives us (a) a SIREN-alternative INR backbone for `BayesianNeuralField`, (b) a principled Bayesian spectral basis where priors over filter parameters correspond to priors over frequency/location atoms, (c) a natural generalization of `RBFFourierFeatures` — one GaborNet layer with zero-mean envelope *is* an RBF-RFF, but stacking them multiplicatively builds an exponentially rich basis the current `pyrox.nn` can't express.

## Background

MFN (Fathony, Sahu, Willmott, Kolter — ICLR 2021) replaces MLP composition with multiplicative filter chaining. Each layer multiplies the previous activation (after an affine) by a new filter `g_i(x)` applied to the *original input*:

$$
z_1 = g_1(x), \qquad
z_{i+1} = g_{i+1}(x) \odot \bigl(W_i z_i + b_i\bigr), \qquad
y = W_L z_L + b_L.
$$

Two filter families, same architecture:

- **FourierNet**: $g_i(x) = \sin(\Omega_i x + \varphi_i)$ — frequency-domain filters. Equivalent to learning a basis of sinusoids whose products (via product-to-sum) span exponentially many frequencies with depth $L$.
- **GaborNet**: $g_i(x) = \sin(\Omega_i x + \varphi_i) \odot \exp\!\bigl(-\tfrac{\gamma_i}{2}\|x - \mu_i\|^2\bigr)$ — Gabor atoms with learned frequency $\Omega_i$, phase $\varphi_i$, location $\mu_i$, and bandwidth $\gamma_i$.

Why this matters for pyrox:

- **Bridges INR and spectral-kernel worlds.** A GaborNet with depth 1 is exactly an RBF random-feature approximation. With depth $L$ it generalizes to a multi-resolution basis whose effective kernel is the Hadamard product of $L$ RBF kernels — unreachable with any existing `pyrox.nn` layer.
- **No init ceremony.** SIREN requires the three-regime `first/hidden/last` init with tuned `c, omega`. MFN works with standard Gaussian / uniform init on all parameters — easier to teach, fewer ways to silently mis-train.
- **Bayesian priors have spectral meaning.** Priors on $\Omega_i, \mu_i$ map directly onto priors over frequency content and spatial localization. That's a research-relevant property the current `BayesianNeuralField` doesn't expose.

### References

- Fathony, Sahu, Willmott, Kolter. **"Multiplicative Filter Networks."** ICLR 2021. [OpenReview](https://openreview.net/forum?id=OmtmcPkkhT)
- Reference impl (authors, PyTorch): [github.com/boschresearch/multiplicative-filter-networks](https://github.com/boschresearch/multiplicative-filter-networks)
- Sitzmann et al. **"SIREN."** NeurIPS 2020. [arXiv:2006.09661](https://arxiv.org/abs/2006.09661) — the competing INR baseline.
- Tancik et al. **"Fourier Features Let Networks Learn High Frequency Functions."** NeurIPS 2020. [arXiv:2006.10739](https://arxiv.org/abs/2006.10739) — direct ancestor of the FourierNet filter.
- Saragadam et al. **"WIRE: Wavelet Implicit Neural Representations."** CVPR 2023. [arXiv:2301.05187](https://arxiv.org/abs/2301.05187) — MFN descendant with complex-Gabor filters (out of scope for this issue, notable for context).

## Proposed API

All public classes land in `src/pyrox/nn/_layers.py` and re-export from `pyrox.nn`.

### Filter primitives

```python
class FourierFilter(PyroxModule):
    """Single Fourier filter: ``g(x) = sin(Ω x + φ)``.

    Parameters:
        Omega: frequency matrix of shape ``(out_features, in_features)``
        phi:   phase vector of shape ``(out_features,)``

    Init follows Fathony et al. 2021 §4.1: ``Ω ~ N(0, σ_f² / in_features)``
    with ``σ_f`` a user-supplied scale (default ``256``), ``φ ~ U(-π, π)``.
    Total param count is ``out_features * (in_features + 1)``.
    """
    Omega: Float[Array, "out in"]
    phi:   Float[Array, " out"]
    in_features:  int
    out_features: int

    @classmethod
    def init(
        cls, in_features: int, out_features: int,
        *, key: PRNGKeyArray, freq_scale: float = 256.0,
        pyrox_name: str | None = None,
    ) -> "FourierFilter": ...

    @pyrox_method
    def __call__(self, x: Float[Array, "N D"]) -> Float[Array, "N H"]: ...


class GaborFilter(PyroxModule):
    """Single Gabor filter: ``g(x) = sin(Ω x + φ) ⊙ exp(-γ/2 ||x - μ||²)``.

    Parameters:
        Omega: frequency matrix ``(out, in)``
        phi:   phase vector ``(out,)``
        mu:    location centers ``(out, in)``
        log_gamma: log-bandwidth ``(out,)``  (parameterized in log space
                   so ``γ > 0`` without a constraint on the optimiser)

    Init follows Fathony et al. 2021 §4.2: ``γ ~ Gamma(α, β)`` with
    defaults ``(α=6, β=1)``; ``μ ~ U(domain_low, domain_high)``;
    ``Ω_i ~ N(0, γ_i · I)`` (frequency scale *tied* to that filter's
    inverse bandwidth, the load-bearing detail from the paper).
    """
    Omega: Float[Array, "out in"]
    phi:   Float[Array, " out"]
    mu:    Float[Array, "out in"]
    log_gamma: Float[Array, " out"]
    in_features: int
    out_features: int
    domain: tuple[float, float]  # static; (low, high) for μ init

    @classmethod
    def init(
        cls, in_features: int, out_features: int,
        *, key: PRNGKeyArray,
        domain: tuple[float, float] = (-1.0, 1.0),
        gamma_alpha: float = 6.0, gamma_beta: float = 1.0,
        pyrox_name: str | None = None,
    ) -> "GaborFilter": ...

    @pyrox_method
    def __call__(self, x: Float[Array, "N D"]) -> Float[Array, "N H"]: ...
```

Both filters are single-layer ops. `FourierFilter` is equivalent to one block of a Fourier feature encoder; `GaborFilter` is equivalent to one block of a *localized* Fourier feature encoder. Users can compose them by hand if they want a custom topology.

### Composite networks

```python
class FourierNet(PyroxModule):
    """Multiplicative Fourier Filter Network.

    Topology (Fathony et al. 2021 eq. 3):
        z_1 = g_1(x)
        z_{i+1} = g_{i+1}(x) * (W_i z_i + b_i)     (i = 1..L-1)
        y = W_L z_L + b_L

    Where each ``g_i`` is a FourierFilter of width ``hidden_features``.
    """
    filters: list[FourierFilter]        # length L
    linears: list[eqx.nn.Linear]        # length L  (last one is the readout)
    in_features: int
    hidden_features: int
    out_features: int
    depth: int                          # L

    @classmethod
    def init(
        cls, in_features: int, hidden_features: int, out_features: int,
        *, depth: int, key: PRNGKeyArray,
        freq_scale: float = 256.0,
        pyrox_name: str | None = None,
    ) -> "FourierNet": ...

    @pyrox_method
    def __call__(self, x: Float[Array, "N D"]) -> Float[Array, "N O"]: ...


class GaborNet(PyroxModule):
    """Multiplicative Gabor Filter Network — same topology as FourierNet,
    but each ``g_i`` is a GaborFilter.
    """
    filters: list[GaborFilter]
    linears: list[eqx.nn.Linear]
    # ... same surface as FourierNet plus `domain`, `gamma_alpha`, `gamma_beta`
```

### Bayesian variants

```python
class BayesianFourierNet(FourierNet):
    """FourierNet with priors on all filter + linear weights.

    Registers sites ``{scope}.filter_{i}.Omega``,
    ``{scope}.filter_{i}.phi``, ``{scope}.linear_{i}.W``,
    ``{scope}.linear_{i}.b`` via ``pyrox_sample``. Prior is
    ``Normal(0, prior_std)`` for Omega/W/b and ``Uniform(-π, π)`` for
    phases. ``prior_std`` is a single scalar; per-parameter scaling is
    out of scope (users can subclass).
    """
    prior_std: float = 1.0


class BayesianGaborNet(GaborNet):
    """GaborNet with priors on filter + linear weights.

    Same convention as BayesianFourierNet, plus priors on ``μ`` (Uniform
    over ``domain``) and ``log_γ`` (Normal(mean, std) in log space,
    defaulting to match the init Gamma in log moment).
    """
    prior_std: float = 1.0
```

Both Bayesian variants are *thin* subclasses — they override `__call__` to register sample sites but share all other state with the deterministic parent. Pattern matches `BayesianFiLM` from issue #86.

### Convenience pure-JAX helper (escape hatch)

```python
def mfn_forward(
    x: Float[Array, "N D"],
    filters: list[Callable[[Array], Array]],
    linears: list[Callable[[Array], Array]],
) -> Float[Array, "N O"]:
    """Pure-JAX MFN forward pass given user-supplied filter and linear callables.

    Exists so users can plug custom filters (e.g. wavelet, complex-Gabor)
    into the MFN topology without subclassing. ``filters`` and ``linears``
    must have the same length L; the forward runs L filter applications
    and L linear applications per the standard MFN recursion.
    """
```

Mirrors the `apply_film` escape hatch — lets researchers try new filter families (WIRE, learnable wavelets) without waiting on upstream support.

## Integration points

- **`BayesianNeuralField`** (bayesnf port): `GaborNet` is a drop-in replacement for the internal MLP. One-line swap, same priors.
- **`BayesianFiLM`** (#86): conditional GaborNet = modulated Gabor atoms — each series gets its own set of frequency/location modulations. Research-interesting composition.
- **`RBFFourierFeatures`**: documented connection in the class docstring — `GaborNet(depth=1, hidden_features=M).filters[0]` is mathematically a single-scale RBF-RFF. Cross-link both ways.
- **`HSGPFeatures`**: same story — `FourierNet(depth=1)` is a learnable-frequency analog of HSGP's fixed Dirichlet basis. Useful comparison in docs.
- **SIREN layer** (planned separate issue): MFN is positioned as the SIREN *alternative*. Both land in `_layers.py`; users pick one based on the INR task.

### File layout

- **Edit** `src/pyrox/nn/_layers.py`: add `FourierFilter`, `GaborFilter`, `FourierNet`, `GaborNet`, `BayesianFourierNet`, `BayesianGaborNet`, `mfn_forward`. Ordering: primitives first, composites second, Bayesian variants third, helper last.
- **Edit** `src/pyrox/nn/__init__.py`: re-export the seven new public names.
- **New** `tests/nn/test_mfn.py`: unit + property + correctness tests (see below).
- **New** `docs/api/nn/mfn.md`: mkdocstrings stubs with the Fathony-2021 topology diagram in the page body and cross-links to `RBFFourierFeatures` / `HSGPFeatures`.
- **Edit** `mkdocs.yml`: nav entry under `nn/`.

### Init (load-bearing — follow Fathony §4)

- `FourierFilter`: `Ω ~ N(0, freq_scale² / in_features)`, `φ ~ U(-π, π)`. **Do not** use the stddev naively — the scaling by `in_features` is what lets the filter resolve high frequencies at init without blowing up.
- `GaborFilter`: the tied scheme is the load-bearing part:
  1. Draw `γ_i ~ Gamma(α=6, β=1)` per filter.
  2. Draw `μ_i ~ U(domain_low, domain_high)` per input dim.
  3. Draw `Ω_i,: ~ N(0, γ_i · I_in)` — note the covariance is the filter-specific `γ_i`, not a shared scale.
- Linear layers `W_i, b_i`: standard equinox `eqx.nn.Linear` default init (Kaiming uniform). No change.

### Forward pass — einops discipline

```python
@pyrox_method
def __call__(self, x):
    # (N, D) -> (N, H)
    z = self.filters[0](x)
    for f, lin in zip(self.filters[1:], self.linears[:-1], strict=True):
        z = f(x) * jax.vmap(lin)(z)    # elementwise product
    return jax.vmap(self.linears[-1])(z)
```

No raw `jnp.einsum`/`reshape` needed — linear layers and elementwise multiplication only. Matches existing layer code style.

### Handling `x.ndim == 1`

Single-point input `(D,)` → `atleast_2d` → run 2D path → squeeze. Same pattern as FiLM (#86) — document in docstring.

### Type annotations

```python
from jaxtyping import Array, Float, PRNGKeyArray
```

Per-file ruff ignore `"src/pyrox/nn/**" = ["F722"]` already in effect. No new config.

## Tests — `tests/nn/test_mfn.py`

All tests use existing numpyro/equinox fixtures for consistency with `test_spectral_layers.py`.

1. **`test_fourier_filter_output_shape`** — `FourierFilter(in=3, out=16)` on `x: (5, 3)` → `(5, 16)`.
2. **`test_gabor_filter_output_shape`** — same, for `GaborFilter`.
3. **`test_gabor_filter_envelope_decays`** — at `x = μ` the envelope is 1.0; at `x` far from `μ` the envelope is ≤ `exp(-γ/2 · large)` → effectively 0. Property test with synthetic `μ, γ` substituted in.
4. **`test_fourier_net_output_shape`** — `FourierNet(in=2, hidden=32, out=4, depth=3)` on `x: (10, 2)` → `(10, 4)`.
5. **`test_gabor_net_output_shape`** — same, for `GaborNet`.
6. **`test_mfn_depth_1_matches_single_filter`** — `FourierNet(depth=1)` forward = `readout(FourierFilter(x))`. Sanity check on the recursion.
7. **`test_gabor_depth_1_matches_rbf_rff_kernel`** — property test with **large `hidden_features`** (e.g. 512): the empirical kernel `GaborNet(depth=1, zero-mean envelope).forward(x) · forward(x')` approximates the RBF kernel `exp(-||x-x'||² / (2ℓ²))` to within Monte-Carlo tolerance. Ties MFN to existing spectral machinery — **this is the load-bearing correctness anchor**.
8. **`test_fourier_net_jits`** — forward under `jax.jit` returns finite values. Mirrors `test_hsgp_jits`.
9. **`test_gabor_net_jits`** — same.
10. **`test_mfn_gradient_flows_through_all_params`** — `jax.grad` of a scalar loss w.r.t. the full network PyTree has finite non-zero gradient on every leaf (filters and linears). Catches accidental `stop_gradient` regressions.
11. **`test_bayesian_fourier_net_registers_sites`** — traced forward under `handlers.trace()` registers exactly the expected set of sample sites (per filter: `Omega`, `phi`; per linear: `W`, `b`). Assert site *count* is `2·L + 2·L = 4L`.
12. **`test_bayesian_gabor_net_registers_sites`** — same, plus `μ` and `log_γ` per filter (6L sites total).
13. **`test_mfn_forward_helper_matches_fourier_net`** — `mfn_forward` given the deconstructed filter/linear callables produces the same output as `FourierNet.__call__`. Validates the escape hatch.
14. **`test_mfn_rejects_mismatched_depth`** — `FourierNet.init(..., depth=0)` raises `ValueError` with message containing `"depth"`.

Keep tests fast — no training loops. The kernel-approximation test (#7) uses large `hidden_features` but only one forward pass per kernel entry → still sub-second.

## Docs

- mkdocstrings page `docs/api/nn/mfn.md`, one `:::` block per public class + helper.
- Page body includes (a) the MFN recursion equation, (b) an explicit callout linking `GaborNet(depth=1)` to `RBFFourierFeatures`, (c) a 6-line usage example chaining a `GaborNet` with an existing likelihood.
- **No notebook** in this issue — defer to the planned `bayesian_neural_fields.py` notebook, which will gain an MFN section in a follow-up.

## Acceptance criteria

- [ ] `pyrox.nn.FourierFilter`, `pyrox.nn.GaborFilter`, `pyrox.nn.FourierNet`, `pyrox.nn.GaborNet`, `pyrox.nn.BayesianFourierNet`, `pyrox.nn.BayesianGaborNet`, `pyrox.nn.mfn_forward` are public and documented.
- [ ] All 14 tests in `tests/nn/test_mfn.py` pass locally and under CI.
- [ ] `uv run pytest -q` green.
- [ ] `uv run --group lint ruff check .` green (entire repo).
- [ ] `uv run --group lint ruff format --check .` green (entire repo).
- [ ] `uv run --group typecheck ty check src/pyrox` green.
- [ ] **`test_gabor_depth_1_matches_rbf_rff_kernel`** passes — this is the non-negotiable correctness check that MFN is correctly implemented. If it fails, the init scheme is wrong.

## Out of scope (explicit)

- **WIRE / complex-Gabor filters.** The `mfn_forward` escape hatch is enough for researchers to try them; promoting to first-class is a separate issue.
- **Learnable filter-family switching** (e.g. mix Fourier + Gabor per layer) — users compose by hand.
- **Amortized / hypernetwork priors** on filter params — this is the FiLM cross-product, out of scope here.
- **MFN-specific pruning or distillation** — separate topic.
- **CUDA kernel fusion** for the multiplicative recursion — pure JAX is plenty fast; premature optimization.
- **Notebook.** Defer to the bayesian_neural_fields notebook follow-up.

## Estimated effort

Medium. Two filter primitives (~40 lines each), two composite networks (~80 lines each), two Bayesian variants (~30 lines each), one pure helper (~15 lines), tests (~250 lines), docs (~60 lines). Single PR, no new dependencies. Main risk: getting the GaborNet init scheme wrong — test #7 is the guardrail.

## Relationships

- Parent (theme epic): #36
- Blocked by: #
- Blocks: #
- Related: #
